inst/atime/tests.R

```diff
@@ -0,0 +1,69 @@
+pkg.edit.fun = function(old.Package, new.Package, sha, new.pkg.path) {
+      pkg_find_replace <- function(glob, FIND, REPLACE) {
+        atime::glob_find_replace(file.path(new.pkg.path, glob), FIND, REPLACE)
+      }
+      Package_regex <- gsub(".", "_?", old.Package, fixed = TRUE)
+      Package_ <- gsub(".", "_", old.Package, fixed = TRUE)
+      new.Package_ <- paste0(Package_, "_", sha)
+      pkg_find_replace(
+        "DESCRIPTION", 
+        paste0("Package:\\s+", old.Package),
+        paste("Package:", new.Package))
+      pkg_find_replace(
+        file.path("src", "Makevars.*in"),
+        Package_regex,
+        new.Package_)
+      pkg_find_replace(
+        file.path("R", "onLoad.R"),
+        Package_regex,
+        new.Package_)
+      pkg_find_replace(
+        file.path("R", "onLoad.R"),
+        sprintf('packageVersion\\("%s"\\)', old.Package),
+        sprintf('packageVersion\\("%s"\\)', new.Package))
+      pkg_find_replace(
+        file.path("src", "init.c"),
+        paste0("R_init_", Package_regex),
+        paste0("R_init_", gsub("[.]", "_", new.Package_)))
+      pkg_find_replace(
+        "NAMESPACE",
+        sprintf('useDynLib\\("?%s"?', Package_regex),
+        paste0('useDynLib(', new.Package_))
+    }
+
+test.list <- list(
+  # Performance regression discussed in: https://github.com/Rdatatable/data.table/issues/4311 
+  # Fixed in: https://github.com/Rdatatable/data.table/pull/4440
+  "Test regression fixed in #4440" = list(
+    pkg.edit.fun = pkg.edit.fun,
+    N = 10^seq(3,8),
+    setup = quote({
+      set.seed(1L)
+      dt <- data.table(a = sample(N, N))
+      setindex(dt, a)
+    }),
+    expr = quote(data.table:::shallow(dt)),
+    "Before" = "9d3b9202fddb980345025a4f6ac451ed26a423be", # This should be changed later. Currently, the source of regression (or the particular commit that led to it) is not clear. In addition, older versions of data.table are having problems when being installed in this manner. (This includes commits from before Mar 20, 2020 or when the issue that discovered or first mentioned the regression was created)
+    "Regression" = "752012f577f8e268bb6d0084ca39a09fa7fbc1c4", # A commit that is affected by the regression: https://github.com/Rdatatable/data.table/commit/752012f577f8e268bb6d0084ca39a09fa7fbc1c4
```

Comment:
this documentation of where the commit ID comes from is not sufficient. 
Right now the comment contains a link to that commit ID, which is redundant with the commit ID itself (does not add any useful information).
We need to know where this commit ID was found. Typically the Regression commit can be the parent of the first commit in the PR which provides the fix, in this case https://github.com/Rdatatable/data.table/pull/4440/commits
The first commit in that PR is https://github.com/Rdatatable/data.table/commit/0f0e7127b880df8459b0ed064dc841acd22f5b73 and on its commit page we can see that the parent is b1b1832b0d2d4032b46477d9fe6efb15006664f4 (not 752012f577f8e268bb6d0084ca39a09fa7fbc1c4 which is the current commit ID in the code, why?).
So a sufficient documentation code/comment would be:
```r
Regression="b1b1832b0d2d4032b46477d9fe6efb15006664f4", # parent of the first commit (https://github.com/Rdatatable/data.table/commit/0f0e7127b880df8459b0ed064dc841acd22f5b73) in the PR (https://github.com/Rdatatable/data.table/pull/4440/commits) that fixes the regression
```

Reply:
please can you properly document all of the other commit IDs, as I suggested above? You may consult https://github.com/DorisAmoakohene/PerformanceTest_data.table since I believe Doris has already done some documentation/comments like this.

Also please remove un-necessary quotes, change "Regression" to Regression in the code. (quotes are not necessary unless the argument names is non-syntactic, meaning starts with a non-letter or contains special characters)

Reply:
> Right now the comment contains a link to that commit ID, which is redundant with the commit ID itself (does not add any useful information). 

I agree, and I thought the same (that it does not add any useful information). That's why I did not have those links at the beginning for my examples. But then yesterday you [wrote](https://github.com/Anirban166/data.table/pull/6#discussion_r1559703661) about including 'links' here, and for the [second](https://github.com/Anirban166/data.table/pull/6#discussion_r1559704932) example you gave it had one such redundant link for 'Before' which made me think that was acceptable for you, or what you wanted. 

I already looked up the two test cases I'm using here on that repository yesterday before committing and they only had those type of links (for commit 'x', `https://github.com/Rdatatable/data.table/commit/x`) for both cases in the comments, and nothing else (Now I see that it got updated three hours again, but that wasn't the case yesterday). I thought I should preserve what Doris did and assumed that you both probably discussed this earlier since I didn't find the same suggestion (to remove such links) on her repository.

> We need to know where this commit ID was found/ how we can verify that it is correct.

Yes, agreed.

> So a sufficient documentation code/comment would be:
```r
Regression="b1b1832b0d2d4032b46477d9fe6efb15006664f4", # parent of the first commit (https://github.com/Rdatatable/data.table/commit/0f0e7127b880df8459b0ed064dc841acd22f5b73) in the PR (https://github.com/Rdatatable/data.table/pull/4440/commits) that fixes the regression
```

Thanks for the example.

> please can you properly document all of the other commit IDs, as I suggested above?
> Also please remove un-necessary quotes, change "Regression" to Regression in the code.

Certainly, I will do that soon (currently debugging some errors while developing my function to get the optimal thread count). I didn't want to change the code for the tests (since that's not what I worked on, or I believe it's Doris' part in discussion with you from earlier) so I left the quotes there but I'll change it now along with the other things you mentioned.

Reply:
> Right now the comment contains a link to that commit ID, which is redundant with the commit ID itself (does not add any useful information). 

I agree, and I thought the same (that it does not add any useful information). That's why I did not have those links at the beginning for my examples. But then yesterday you [wrote](https://github.com/Anirban166/data.table/pull/6#discussion_r1559703661) about including 'links' here, and for the [second](https://github.com/Anirban166/data.table/pull/6#discussion_r1559704932) example you gave it had one such redundant link for 'Before' which made me think that was acceptable for you, or what you wanted. 

I already looked up the two test cases I'm using here on that repository yesterday before committing and they only had those type of links (for commit 'x', `https://github.com/Rdatatable/data.table/commit/x`) for both cases in the comments, and nothing else (now I see that it got updated four hours back, but that wasn't the case yesterday). I thought I should preserve what Doris did and assumed that you both probably discussed this earlier since I didn't find the same suggestion (to remove such links) on her repository.

> We need to know where this commit ID was found/ how we can verify that it is correct.

Yes, agreed.

> So a sufficient documentation code/comment would be:
```r
Regression="b1b1832b0d2d4032b46477d9fe6efb15006664f4", # parent of the first commit (https://github.com/Rdatatable/data.table/commit/0f0e7127b880df8459b0ed064dc841acd22f5b73) in the PR (https://github.com/Rdatatable/data.table/pull/4440/commits) that fixes the regression
```

Thanks for the example.

> please can you properly document all of the other commit IDs, as I suggested above?
> Also please remove un-necessary quotes, change "Regression" to Regression in the code.

Certainly, I will do that shortly (currently debugging some errors while developing my function to get the optimal thread count). I didn't want to change the code for the tests (since that's not what I worked on, or I believe it's Doris' part in discussion with you from earlier) so I left the quotes there but I'll change it soon now along with the other things you mentioned.

Reply:
> But then yesterday you https://github.com/Anirban166/data.table/pull/6#discussion_r1559703661 about including 'links' here, and for the https://github.com/Anirban166/data.table/pull/6#discussion_r1559704932 example you gave it had one such redundant link for 'Before' which made me think that was acceptable for you, or what you wanted.

Never mind this, I reread and saw the correct changes right now - my bad! (The new commits made me think I was looking at another test case)

Reply:
> But then yesterday you https://github.com/Anirban166/data.table/pull/6#discussion_r1559703661 about including 'links' here, and for the https://github.com/Anirban166/data.table/pull/6#discussion_r1559704932 example you gave it had one such redundant link for 'Before' which made me think that was acceptable for you, or what you wanted.

Never mind this, I just reread and saw the corrected changes right now - my bad! (The new commits made me think I was looking at another test case)